File: aisongtool/ace_step_api.py
"""HTTP client for ACE-Step-1.5's own `acestep-api` REST server
(https://github.com/ACE-Step/ACE-Step-1.5) — replaces the earlier client for
acestep.cpp's `ace-server` (a two-stage `/lm` + `/synth` + `/job?id=` job
protocol). ACE-Step-1.5's API is a single async job queue instead:
`POST /release_task` submits one task (caption+lyrics -> a finished song in
one step), `POST /query_result` polls it by `task_id`, and the finished file
is fetched via a plain `GET /v1/audio?path=...` download.

Mirrors `desktop/src/main/tools/ace-step-api.ts` — see that file's docstring
for the same protocol explanation. This Python copy backs the Flet app's
Docker-only "Song Generation" view (see `flet_app/views/_create_pipeline.py`)
and has not itself been exercised against a live server yet, same caveat as
the TS client it mirrors.
"""
from __future__ import annotations

import time
from pathlib import Path
from typing import Callable

import httpx

from .ace_step import DEFAULT_DIT_MODEL

LogFn = Callable[[str], None]

DEFAULT_HOST = "127.0.0.1"
DEFAULT_PORT = 8001


class AceStepApiError(RuntimeError):
    pass


def _base_url(host: str, port: int) -> str:
    return f"http://{host}:{port}"


def is_server_up(host: str = DEFAULT_HOST, port: int = DEFAULT_PORT) -> bool:
    try:
        resp = httpx.get(f"{_base_url(host, port)}/health", timeout=3.0)
        return resp.status_code == 200
    except httpx.HTTPError:
        return False


def wait_for_server(
    host: str = DEFAULT_HOST,
    port: int = DEFAULT_PORT,
    timeout: float = 300.0,
    log: LogFn = print,
) -> bool:
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if is_server_up(host, port):
            return True
        time.sleep(2.0)
    log(f"ACE-Step server did not become healthy within {timeout:.0f}s.")
    return False


def generate_song(
    prompt: str,
    lyrics: str,
    duration: float,
    out_dir: Path,
    log: LogFn = print,
    host: str = DEFAULT_HOST,
    port: int = DEFAULT_PORT,
    model: str = DEFAULT_DIT_MODEL,
    vocal_language: str = "en",
    instrumental: bool = False,
    seed: int | None = None,
    poll_interval: float = 2.0,
    timeout: float = 600.0,
    on_progress: LogFn | None = None,
) -> Path:
    """Submit one `/release_task` (caption+lyrics -> a finished song in a
    single job, no separate codes/audio stages like acestep.cpp's
    `/lm`+`/synth`), poll `/query_result` until done, then download the
    finished file from `/v1/audio?path=...`.

    `prompt`/`lyrics` are always treated as literal — the caller has already
    resolved who wrote them (manual text or Gemma 4). `instrumental=True`
    sends empty lyrics (no separate "[Instrumental]" convention confirmed for
    this API, unlike acestep.cpp's documented one). `vocal_language` should
    already be resolved to a real code rather than left as "unknown"."""
    base = _base_url(host, port)
    task_body: dict = {
        "task_type": "text2music",
        "prompt": prompt,
        "lyrics": "" if instrumental else lyrics,
        "audio_duration": duration,
        "vocal_language": "en" if vocal_language == "unknown" else vocal_language,
        "seed": -1 if seed is None else seed,
        "batch_size": 1,
        "model": model,
    }

    log(f"POST {base}/release_task {task_body}")
    try:
        with httpx.Client(timeout=30.0) as client:
            task_id = _release_task(client, base, task_body, log)
            result = _poll_result(client, base, task_id, log, poll_interval, timeout, on_progress)
            file_path = result.get("file")
            if not file_path:
                raise AceStepApiError(f"task {task_id} succeeded but returned no file path: {result}")

            log(f"GET {base}{file_path}")
            audio_resp = client.get(f"{base}{file_path}", timeout=120.0)
            if audio_resp.status_code != 200:
                raise AceStepApiError(f"Downloading result audio failed ({audio_resp.status_code})")
            audio_bytes = audio_resp.content
    except Exception as exc:  # noqa: BLE001
        log(f"Generation failed: {exc!r}")
        raise

    out_dir.mkdir(parents=True, exist_ok=True)
    out_path = out_dir / "generated.mp3"
    out_path.write_bytes(audio_bytes)
    log(f"Saved generated song to {out_path}")
    return out_path


def _release_task(client: httpx.Client, base: str, body: dict, log: LogFn) -> str:
    resp = client.post(f"{base}/release_task", json=body, timeout=30.0)
    if resp.status_code != 200:
        raise AceStepApiError(f"/release_task failed ({resp.status_code}): {resp.text}")
    task_id = (resp.json() or {}).get("data", {}).get("task_id")
    if not task_id:
        raise AceStepApiError(f"/release_task returned no task_id: {resp.text}")
    log(f"task submitted: {task_id}")
    return task_id


def _poll_result(
    client: httpx.Client, base: str, task_id: str, log: LogFn, poll_interval: float, timeout: float,
    on_progress: LogFn | None,
) -> dict:
    import json as _json

    deadline = time.monotonic() + timeout
    last_status = None
    while time.monotonic() < deadline:
        time.sleep(poll_interval)
        try:
            resp = client.post(f"{base}/query_result", json={"task_id_list": [task_id]}, timeout=30.0)
        except httpx.HTTPError as exc:
            log(f"Poll request failed ({exc!r}) — server may still be loading models, retrying...")
            continue
        if resp.status_code != 200:
            log(f"/query_result returned {resp.status_code}, retrying...")
            continue
        items = (resp.json() or {}).get("data", [])
        item = next((i for i in items if i.get("task_id") == task_id), None)
        if item is None:
            log(f"/query_result didn't include task {task_id} yet, retrying...")
            continue
        status = item.get("status")
        if status != last_status:
            label = {0: "running", 1: "succeeded", 2: "failed"}.get(status, str(status))
            log(f"task {task_id}: {label}")
            last_status = status
            if on_progress:
                on_progress(f"generating: {label}")
        if status == 1:
            # `result` decodes to an array (one entry per `batch_size`, even
            # when batch_size is 1 — confirmed against a real server
            # response: `[{"file": ..., "metas": ..., ...}]`, not a flat
            # object), mirroring the same fix in ace-step-api.ts.
            result = item.get("result")
            if not result:
                return {}
            parsed = _json.loads(result)
            return parsed[0] if parsed else {}
        if status == 2:
            raise AceStepApiError(f"task {task_id} failed: {item.get('result') or '(no detail)'}")
    raise AceStepApiError(f"task {task_id} timed out after {timeout:.0f}s.")

File: desktop/src/main/tools/ace-step-api.ts
/**
 * HTTP client for ACE-Step-1.5's own `acestep-api` REST server
 * (https://github.com/ACE-Step/ACE-Step-1.5). Its API is a single async job
 * queue: `POST /release_task` submits one task (description -> a finished
 * song in one step, no separate codes-then-audio stages), `POST
 * /query_result` polls it by `task_id`, and the finished file is fetched
 * via a plain `GET /v1/audio?path=...` download.
 */
import { mkdirSync, writeFileSync } from 'fs'
import path from 'path'

export type LogFn = (line: string) => void

const DEFAULT_HOST = '127.0.0.1'
const DEFAULT_PORT = 8001

export class AceStepApiError extends Error {}

function baseUrl(host: string, port: number): string {
  return `http://${host}:${port}`
}

export async function isServerUp(host = DEFAULT_HOST, port = DEFAULT_PORT): Promise<boolean> {
  try {
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), 3000)
    const resp = await fetch(`${baseUrl(host, port)}/health`, { signal: controller.signal })
    clearTimeout(timer)
    return resp.status === 200
  } catch {
    return false
  }
}

/** Poll /health until it responds or `timeoutMs` passes (model loading on
 * first start can take a while). */
export async function waitForServer(
  host = DEFAULT_HOST,
  port = DEFAULT_PORT,
  timeoutMs = 300_000,
  log: LogFn = console.log
): Promise<boolean> {
  const deadline = Date.now() + timeoutMs
  while (Date.now() < deadline) {
    if (await isServerUp(host, port)) return true
    await sleep(2000)
  }
  log(`ACE-Step server did not become healthy within ${Math.round(timeoutMs / 1000)}s.`)
  return false
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export interface GenerateSampleSongOptions {
  /** A short natural-language description — ACE-Step's own sample mode
   * (`sample_mode`/`sample_query`) auto-generates the caption, lyrics, and
   * every other metadata field (bpm/key/etc.) from this via its 5Hz LM, so
   * the caller never has to supply any of that directly. */
  description: string
  model: string
  outDir: string
  log?: LogFn
  host?: string
  port?: number
  pollIntervalMs?: number
  timeoutMs?: number
  onProgress?: LogFn
}

export interface GeneratedSample {
  audioPath: string
  /** The actual lyrics the LM generated for this song, straight from the
   * server's response — needed for subtitle/caption alignment downstream,
   * since the caller never wrote any lyrics itself in sample mode. */
  lyrics: string
}

interface ReleaseTaskResponse {
  data?: { task_id?: string }
  code?: number
  error?: string | null
}

interface QueryResultItem {
  task_id: string
  status: number // 0 = queued/running, 1 = succeeded, 2 = failed
  result?: string // JSON string: { file, lyrics, metas, ... }
}

interface ParsedResultItem {
  file?: string
  lyrics?: string
}

/** Submit one `/release_task` in sample mode, poll `/query_result` until
 * done, then download the finished file from `/v1/audio?path=...`. */
export async function generateSampleSong(opts: GenerateSampleSongOptions): Promise<GeneratedSample> {
  const {
    description,
    model,
    outDir,
    log = console.log,
    host = DEFAULT_HOST,
    port = DEFAULT_PORT,
    pollIntervalMs = 2000,
    timeoutMs = 600_000,
    onProgress
  } = opts

  const base = baseUrl(host, port)
  const taskBody = {
    sample_mode: true,
    sample_query: description,
    thinking: true,
    batch_size: 1,
    model
  }

  log(`POST ${base}/release_task ${JSON.stringify(taskBody)}\r\n`)
  try {
    const taskId = await releaseTask(base, taskBody, log)
    const resultJson = await pollResult(base, taskId, log, pollIntervalMs, timeoutMs, onProgress)
    const filePath = resultJson.file
    if (!filePath) {
      throw new AceStepApiError(`Task ${taskId} succeeded but returned no file path: ${JSON.stringify(resultJson)}`)
    }

    log(`GET ${base}${filePath}\r\n`)
    const audioResp = await fetch(`${base}${filePath}`)
    if (!audioResp.ok) {
      throw new AceStepApiError(`Downloading result audio failed (${audioResp.status})`)
    }
    const audioBuffer = Buffer.from(await audioResp.arrayBuffer())

    mkdirSync(outDir, { recursive: true })
    const audioPath = path.join(outDir, 'generated.mp3')
    writeFileSync(audioPath, audioBuffer)
    log(`Saved generated song to ${audioPath}\r\n`)
    return { audioPath, lyrics: resultJson.lyrics ?? '' }
  } catch (exc) {
    log(`Generation failed: ${String(exc)}\r\n`)
    throw exc
  }
}

async function releaseTask(base: string, body: Record<string, unknown>, log: LogFn): Promise<string> {
  const resp = await fetch(`${base}/release_task`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  })
  const text = await resp.text()
  if (!resp.ok) {
    throw new AceStepApiError(`/release_task failed (${resp.status}): ${text}`)
  }
  const parsed = JSON.parse(text) as ReleaseTaskResponse
  const taskId = parsed.data?.task_id
  if (!taskId) throw new AceStepApiError(`/release_task returned no task_id: ${text}`)
  log(`task submitted: ${taskId}\r\n`)
  return taskId
}

async function pollResult(
  base: string,
  taskId: string,
  log: LogFn,
  pollIntervalMs: number,
  timeoutMs: number,
  onProgress: LogFn | undefined
): Promise<ParsedResultItem> {
  const deadline = Date.now() + timeoutMs
  let lastStatus: number | null = null
  while (Date.now() < deadline) {
    await sleep(pollIntervalMs)
    let resp: Response
    try {
      resp = await fetch(`${base}/query_result`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ task_id_list: [taskId] })
      })
    } catch (exc) {
      log(`Poll request failed (${String(exc)}) — server may still be loading models, retrying...\r\n`)
      continue
    }
    if (resp.status !== 200) {
      log(`/query_result returned ${resp.status}, retrying...\r\n`)
      continue
    }
    const payload = (await resp.json()) as { data?: QueryResultItem[] }
    const item = payload.data?.find((i) => i.task_id === taskId)
    if (!item) {
      log(`/query_result didn't include task ${taskId} yet, retrying...\r\n`)
      continue
    }
    if (item.status !== lastStatus) {
      const label = item.status === 0 ? 'running' : item.status === 1 ? 'succeeded' : 'failed'
      log(`task ${taskId}: ${label}\r\n`)
      lastStatus = item.status
      onProgress?.(`generating: ${label}`)
    }
    if (item.status === 1) {
      // `result` decodes to an array (one entry per `batch_size`, even when
      // batch_size is 1 — confirmed against a real server response, which
      // came back as `[{file, metas, ...}]`, not a flat object).
      if (!item.result) return {}
      const parsed = JSON.parse(item.result) as ParsedResultItem[]
      return parsed[0] ?? {}
    }
    if (item.status === 2) {
      throw new AceStepApiError(`task ${taskId} failed: ${item.result ?? '(no detail)'}`)
    }
  }
  throw new AceStepApiError(`task ${taskId} timed out after ${Math.round(timeoutMs / 1000)}s.`)
}
